egs/librispeech/ASR/transducer_stateless/beam_search.py

```diff
@@ -111,110 +237,98 @@ def beam_search(
-            # log_prob is (1, 1, 1, vocab_size)
-            log_prob = log_prob.squeeze()
-            # Now log_prob is (vocab_size,)
+                # TODO(fangjun): Ccale the blank posterior
 
-            # If we choose blank here, add the new hypothesis to B.
-            # Otherwise, add the new hypothesis to A
+                log_prob = logits.log_softmax(dim=-1)
+                # log_prob is (1, 1, 1, vocab_size)
+                log_prob = log_prob.squeeze()
+                # Now log_prob is (vocab_size,)
+                joint_cache[cached_key] = log_prob
+            else:
+                log_prob = joint_cache[cached_key]
 
-            # First, choose blank
+            # First, process the blank symbol
             skip_log_prob = log_prob[blank_id]
             new_y_star_log_prob = y_star.log_prob + skip_log_prob.item()
 
             # ys[:] returns a copy of ys
-            new_y_star = Hypothesis(
-                ys=y_star.ys[:],
-                log_prob=new_y_star_log_prob,
-                # Caution: Use y_star.decoder_state here
-                decoder_state=y_star.decoder_state,
-            )
-            B.append(new_y_star)
+            B.add(Hypothesis(ys=y_star.ys[:], log_prob=new_y_star_log_prob))
 
-            # Second, choose other labels
-            for i, v in enumerate(log_prob.tolist()):
-                if i in (blank_id, sos_id):
+            # Second, process other non-blank labels
+            values, indices = log_prob.topk(beam + 1)
+            for i, v in zip(indices.tolist(), values.tolist()):
+                if i == blank_id:
                     continue
                 new_ys = y_star.ys + [i]
                 new_log_prob = y_star.log_prob + v
-                new_hyp = Hypothesis(
-                    ys=new_ys,
-                    log_prob=new_log_prob,
-                    decoder_state=decoder_state,
-                )
-                A.append(new_hyp)
-            u += 1
-            # check whether B contains more than "beam" elements more probable
+                A.add(Hypothesis(ys=new_ys, log_prob=new_log_prob))
+
+            # Check whether B contains more than "beam" elements more probable
             # than the most probable in A
-            A_most_probable = max(A, key=lambda hyp: hyp.log_prob)
-            B = sorted(
-                [hyp for hyp in B if hyp.log_prob > A_most_probable.log_prob],
-                key=lambda hyp: hyp.log_prob,
-                reverse=True,
-            )
-            if len(B) >= beam:
-                B = B[:beam]
+            A_most_probable = A.get_most_probable()
+
+            kept_B = B.filter(A_most_probable.log_prob)
+
+            if len(kept_B) >= beam:
+                B = kept_B.topk(beam)
                 break
+
         t += 1
-    best_hyp = max(B, key=lambda hyp: hyp.log_prob / len(hyp.ys[1:]))
-    ys = best_hyp.ys[1:]  # [1:] to remove the blank
+
+    best_hyp = B.get_most_probable(length_norm=True)
```

Comment:
@danpovey 

It is using length normalized log-prob only at the end to select the best hypothesis, while the total log-prob is used during searching.

That is from the original RNN-T paper: https://arxiv.org/pdf/1211.3711.pdf
> The length normalisation in the final line appears
to be important for good performance, as otherwise
shorter output sequences are excessively favoured over
longer ones; similar techniques are employed for hidden Markov models in speech and handwriting recognition (Bertolami et al., 2006).

Looks like others also follow this approach.

---

I have  tried two cases:
(1) Remove length normalization at the end. It produces the same WER.
```
[test-clean-beam_4] %WER 2.92% [1536 / 52576, 163 ins, 128 del, 1245 sub ]
[test-other-beam_4] %WER 7.37% [3859 / 52343, 428 ins, 334 del, 3097 sub ]
```

(2) Use length normalization also in the search process. It produces slightly worse WER.
```
[test-clean-beam_4] %WER 2.93% [1538 / 52576, 164 ins, 128 del, 1246 sub ]
[test-other-beam_4] %WER 7.38% [3861 / 52343, 433 ins, 330 del, 3098 sub ]
```

Reply:
@csukuangfj FYI, from [Graves et al., 2013](https://arxiv.org/pdf/1303.5778.pdf) Sec 3.3:

> Note the n-best list from the transducer was originally sorted by the length normalised log-probabilty log Pr(y) / |y| ; in the current work we dispense with the normalisation (which only helps when there are many more deletions than insertions) and sort by Pr(y)

Reply:
I see. Thanks for the link.